Importing libraries


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px


Load dataset


In [2]:
df = pd.read_csv('/content/customers-ed-never-charge-last365d.csv-1762179391910.csv')

Explo


In [41]:
df.head(15)

VEHICLE_ID                          MODEL_ID  \
0   1d293c750cc5165af5c7266195b1dac0  7d3278bd8d29df4dbb5dae11ed4e69bd   
1   ca1c90897ce1f12c092fe96acfc8832c  7d3278bd8d29df4dbb5dae11ed4e69bd   
2   ddf0231bd86205eefbb6fde99b1ab756  7d3278bd8d29df4dbb5dae11ed4e69bd   
3   981fb8da358abde5fc063be80e011720  7d3278bd8d29df4dbb5dae11ed4e69bd   
4   981fb8da358abde5fc063be80e011720  7d3278bd8d29df4dbb5dae11ed4e69bd   
5   c71ef46b95f70c617b399ac51b6e846f  7d3278bd8d29df4dbb5dae11ed4e69bd   
6   e60b281a0143cacff6cb7abf453d8dc3  7d3278bd8d29df4dbb5dae11ed4e69bd   
7   1d293c750cc5165af5c7266195b1dac0  7d3278bd8d29df4dbb5dae11ed4e69bd   
8   c6d81f7811636647277b602e831ec5c3  7d3278bd8d29df4dbb5dae11ed4e69bd   
9   d7777da68b50ace0b909872d611db978  7d3278bd8d29df4dbb5dae11ed4e69bd   
10  73f583278b1fb9f19621c240df5b31c5  7d3278bd8d29df4dbb5dae11ed4e69bd   
11  42ac25c2d95a476f0186fa454c3def84  7d3278bd8d29df4dbb5dae11ed4e69bd   
12  3dcaa2ca32c9a3b8cdc566b259f754f0  7d3278bd8d29df4dbb5dae11ed4e69bd   
13  af77c8463ac27b3776ae4528d58704dc  7d3278bd8d29df4dbb5dae11ed4e69bd   
14  f94f58e24061c6c82ec4f0787951c798  7d3278bd8d29df4dbb5dae11ed4e69bd   

               STARTED_TIME            FINISHED_TIME  CHARGELEVELSTART  \
0   2023-01-15 05:46:16.972  2023-01-15 06:06:14.466              72.0   
1   2023-01-15 08:12:02.460  2023-01-15 08:17:27.798               NaN   
2   2023-01-15 10:19:51.668  2023-01-15 10:36:08.477              60.0   
3   2023-01-15 12:57:37.334  2023-01-15 12:59:16.264               NaN   
4   2023-01-15 13:00:56.392  2023-01-15 13:04:52.894               NaN   
5   2023-01-17 02:41:12.222  2023-01-17 02:52:47.979              63.0   
6   2023-01-17 02:47:23.724  2023-01-17 02:59:38.214              65.0   
7   2023-01-17 06:12:38.637  2023-01-17 06:23:35.156              58.0   
8   2023-01-17 06:07:54.693  2023-01-17 06:28:49.737              96.0   
9   2023-01-17 06:00:49.489  2023-01-17 06:33:44.534              71.0   
10  2023-01-17 07:22:00.502  2023-01-17 07:44:56.008              32.0   
11  2023-01-17 07:10:10.330  2023-01-17 07:48:16.569              73.0   
12  2023-01-17 09:35:05.107  2023-01-17 09:36:39.852               NaN   
13  2023-01-17 09:27:20.884  2023-01-17 09:39:00.763              70.0   
14  2023-01-17 09:37:51.703  2023-01-17 09:40:42.858              55.0   

    CHARGELEVELEND CHARGED  SERVICERENTAL  
0             65.0   False          False  
1             65.0   False           True  
2             46.0   False          False  
3            100.0   False           True  
4             98.0   False           True  
5             56.0   False          False  
6             38.0   False          False  
7             52.0   False          False  
8             76.0   False          False  
9             59.0   False          False  
10            16.0   False          False  
11            50.0   False          False  
12             3.0    True           True  
13            62.0   False          False  
14            52.0   False          False

In [4]:
df.describe()


CHARGELEVELSTART  CHARGELEVELEND
count      40149.000000    45661.000000
mean          62.680739       55.973807
std           24.288224       27.110900
min            0.000000        1.000000
25%           42.000000       33.000000
50%           63.000000       55.000000
75%           85.000000       80.000000
max          100.000000      100.000000

In [6]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45919 entries, 0 to 45918
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   VEHICLE_ID        45919 non-null  object 
 1   MODEL_ID          45919 non-null  object 
 2   STARTED_TIME      45919 non-null  object 
 3   FINISHED_TIME     45919 non-null  object 
 4   CHARGELEVELSTART  40149 non-null  float64
 5   CHARGELEVELEND    45661 non-null  float64
 6   CHARGED           45907 non-null  object 
 7   SERVICERENTAL     45919 non-null  bool   
dtypes: bool(1), float64(2), object(5)
memory usage: 2.5+ MB


In [7]:
df.VEHICLE_ID.value_counts()

VEHICLE_ID
99a39a0fcec88de0c6c067adb64a3bb6    1098
12700f3e43ce275c51956087a0f60e7e    1094
0ef2ca2d596258de2677980732d313e3    1075
2db363ebf1e6be2201eb12230dc83b41    1060
42ac25c2d95a476f0186fa454c3def84    1059
ddf0231bd86205eefbb6fde99b1ab756    1056
af77c8463ac27b3776ae4528d58704dc    1053
e60b281a0143cacff6cb7abf453d8dc3    1049
1c347a2ffbbdfb70197160e8e77c9a49    1044
ca1c90897ce1f12c092fe96acfc8832c    1043
3edacefb15e7dbd9c470616287fb15ed    1031
fe24aa75d8c4cfa99df9e08546308ed6    1031
b25e0405b92f09d629d57b36cc03af8e    1027
0819698514bc1e0ddaaad0522259d4dd     993
5435abe93f9567f656340f9639ebe1df     985
3dcaa2ca32c9a3b8cdc566b259f754f0     950
8ec9ae2615eaee29cad7d6faaf8eeb51     944
83d029b4fee4280c9ba965b2a21edfd9     940
6fae9073359b122756c49851f24798d5     939
6b09e23b1b8e95a670127e57b8580d98     935
6a40f9c6a146ffdfa0a820cde2662af8     922
0c5d1b8f3dbe2f559fd3b70a6a31a0ce     912
9cf9888f491b2ee122be311428f3ae4a     894
867d31a17ce4fbc72e3e3f224f717e79     888
8bf2cf4a2196a436618aa24ac35af4b8     887
1d293c750cc5165af5c7266195b1dac0     886
60af77191a7c0571643488e52189428f     881
788a5bb1fe15663febd379c0a835d79f     875
c71ef46b95f70c617b399ac51b6e846f     856
308330becfa60294f45d93d1d359cc15     852
c6d81f7811636647277b602e831ec5c3     848
1e3398251d2ef4394fdfd399bff97c56     845
93642b89382f93989fb67f258a827fb4     844
f94f58e24061c6c82ec4f0787951c798     838
ad5a12e5c0e12dea3c435931a78fa540     838
73f583278b1fb9f19621c240df5b31c5     834
d85d782906d7062168d02c28cdaca753     819
ad657d9e2b87df0576cb0f73e43a8049     816
0592a3e4a455ba32227b24f88215cce6     812
8434d49aa0baa7dad7a1df72defa440a     809
61bedc330d875425f1eb7937a694cc21     804
545e5c24d652895cb04d701b199ae6cf     804
59de72ba2a06fd60b5b7a1cf8827198a     785
80ccf8ddbd6c8eb8a6dd943f7837fb80     770
1ad57f261423114ec397fc6bcf156335     760
b0021abffae86782e17f773e74fbe4a3     755
7218b9c79dc23f8ca0aaff4fd88147db     744
18a86bed65d052f506e5dedcda224325     732
d7777da68b50ace0b909872d611db978     713
981fb8da358abde5fc063be80e011720     702
1a74a25866a91c2036c39351e7dc8c2d     609
dc753ee97d00780d5b36379e37ee78bf     179
Name: count, dtype: int64

In [8]:
df.isna().sum()


VEHICLE_ID             0
MODEL_ID               0
STARTED_TIME           0
FINISHED_TIME          0
CHARGELEVELSTART    5770
CHARGELEVELEND       258
CHARGED               12
SERVICERENTAL          0
dtype: int64

**Cleaning data : duplicated values**


In [9]:
df.duplicated().sum()

np.int64(218)

In [15]:
df_cleaned = df.drop_duplicates()
df_cleaned

VEHICLE_ID                          MODEL_ID  \
0      1d293c750cc5165af5c7266195b1dac0  7d3278bd8d29df4dbb5dae11ed4e69bd   
1      ca1c90897ce1f12c092fe96acfc8832c  7d3278bd8d29df4dbb5dae11ed4e69bd   
2      ddf0231bd86205eefbb6fde99b1ab756  7d3278bd8d29df4dbb5dae11ed4e69bd   
3      981fb8da358abde5fc063be80e011720  7d3278bd8d29df4dbb5dae11ed4e69bd   
4      981fb8da358abde5fc063be80e011720  7d3278bd8d29df4dbb5dae11ed4e69bd   
...                                 ...                               ...   
45914  9cf9888f491b2ee122be311428f3ae4a  7d3278bd8d29df4dbb5dae11ed4e69bd   
45915  0819698514bc1e0ddaaad0522259d4dd  7d3278bd8d29df4dbb5dae11ed4e69bd   
45916  42ac25c2d95a476f0186fa454c3def84  7d3278bd8d29df4dbb5dae11ed4e69bd   
45917  fe24aa75d8c4cfa99df9e08546308ed6  7d3278bd8d29df4dbb5dae11ed4e69bd   
45918  0819698514bc1e0ddaaad0522259d4dd  7d3278bd8d29df4dbb5dae11ed4e69bd   

                  STARTED_TIME            FINISHED_TIME  CHARGELEVELSTART  \
0      2023-01-15 05:46:16.972  2023-01-15 06:06:14.466              72.0   
1      2023-01-15 08:12:02.460  2023-01-15 08:17:27.798               NaN   
2      2023-01-15 10:19:51.668  2023-01-15 10:36:08.477              60.0   
3      2023-01-15 12:57:37.334  2023-01-15 12:59:16.264               NaN   
4      2023-01-15 13:00:56.392  2023-01-15 13:04:52.894               NaN   
...                        ...                      ...               ...   
45914  2023-09-10 01:11:54.272  2023-09-10 01:28:32.977               NaN   
45915  2023-09-11 05:01:33.619  2023-09-15 11:05:58.017              28.0   
45916  2023-09-05 14:52:57.388  2023-09-05 15:07:25.140              77.0   
45917  2023-09-05 16:25:29.189  2023-09-05 16:42:21.229              64.0   
45918  2023-09-05 23:02:29.659  2023-09-05 23:21:07.119              51.0   

       CHARGELEVELEND CHARGED  SERVICERENTAL  
0                65.0   False          False  
1                65.0   False           True  
2                46.0   False          False  
3               100.0   False           True  
4                98.0   False           True  
...               ...     ...            ...  
45914            98.0   False           True  
45915            52.0    True          False  
45916            70.0    True          False  
45917            63.0    True          False  
45918            36.0    True          False  

[45701 rows x 8 columns]

**Cleaning data : Casting types**

In [11]:
df_cleaned.dtypes


VEHICLE_ID           object
MODEL_ID             object
STARTED_TIME         object
FINISHED_TIME        object
CHARGELEVELSTART    float64
CHARGELEVELEND      float64
CHARGED              object
SERVICERENTAL          bool
dtype: object

In [17]:
df_cleaned['CHARGED'] = df_cleaned['CHARGED'].astype('boolean')
df_cleaned['STARTED_TIME'] = pd.to_datetime(df_cleaned['STARTED_TIME'])
df_cleaned['FINISHED_TIME'] = pd.to_datetime(df_cleaned['FINISHED_TIME'])
df_cleaned.dtypes

/tmp/ipython-input-942917857.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['CHARGED'] = df_cleaned['CHARGED'].astype('boolean')
/tmp/ipython-input-942917857.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['STARTED_TIME'] = pd.to_datetime(df_cleaned['STARTED_TIME'])
/tmp/ipython-input-942917857.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documen

VEHICLE_ID                  object
MODEL_ID                    object
STARTED_TIME        datetime64[ns]
FINISHED_TIME       datetime64[ns]
CHARGELEVELSTART           float64
CHARGELEVELEND             float64
CHARGED                    boolean
SERVICERENTAL                 bool
dtype: object

In [13]:
df_cleaned.head()

VEHICLE_ID                          MODEL_ID  \
0  1d293c750cc5165af5c7266195b1dac0  7d3278bd8d29df4dbb5dae11ed4e69bd   
1  ca1c90897ce1f12c092fe96acfc8832c  7d3278bd8d29df4dbb5dae11ed4e69bd   
2  ddf0231bd86205eefbb6fde99b1ab756  7d3278bd8d29df4dbb5dae11ed4e69bd   
3  981fb8da358abde5fc063be80e011720  7d3278bd8d29df4dbb5dae11ed4e69bd   
4  981fb8da358abde5fc063be80e011720  7d3278bd8d29df4dbb5dae11ed4e69bd   

             STARTED_TIME           FINISHED_TIME  CHARGELEVELSTART  \
0 2023-01-15 05:46:16.972 2023-01-15 06:06:14.466              72.0   
1 2023-01-15 08:12:02.460 2023-01-15 08:17:27.798               NaN   
2 2023-01-15 10:19:51.668 2023-01-15 10:36:08.477              60.0   
3 2023-01-15 12:57:37.334 2023-01-15 12:59:16.264               NaN   
4 2023-01-15 13:00:56.392 2023-01-15 13:04:52.894               NaN   

   CHARGELEVELEND  CHARGED  SERVICERENTAL  
0            65.0    False          False  
1            65.0    False           True  
2            46.0    False          False  
3           100.0    False           True  
4            98.0    False           True

**Adding duration data for future analysis**

In [24]:
df_cleaned['DURATION'] = df_cleaned['FINISHED_TIME'] - df_cleaned['STARTED_TIME']

/tmp/ipython-input-3029558976.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['DURATION'] = df_cleaned['FINISHED_TIME'] - df_cleaned['STARTED_TIME']


In [25]:
df_cleaned['DURATION']

0       0 days 00:19:57.494000
1       0 days 00:05:25.338000
2       0 days 00:16:16.809000
3       0 days 00:01:38.930000
4       0 days 00:03:56.502000
                 ...          
45914   0 days 00:16:38.705000
45915   4 days 06:04:24.398000
45916   0 days 00:14:27.752000
45917   0 days 00:16:52.040000
45918   0 days 00:18:37.460000
Name: DURATION, Length: 45701, dtype: timedelta64[ns]

In [26]:
df_cleaned['DURATION_MIN'] = df_cleaned['DURATION'].dt.total_seconds() / 60

/tmp/ipython-input-1472647608.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['DURATION_MIN'] = df_cleaned['DURATION'].dt.total_seconds() / 60


**Cleaning Data : handling null values**

In [35]:
df_cleaned.isna().sum()

VEHICLE_ID             0
MODEL_ID               0
STARTED_TIME           0
FINISHED_TIME          0
CHARGELEVELSTART    5749
CHARGELEVELEND       258
CHARGED               12
SERVICERENTAL          0
DURATION               0
DURATION_MIN           0
dtype: int64

In [47]:
df_cleaned[df_cleaned['CHARGELEVELSTART'].isna()]

VEHICLE_ID                          MODEL_ID  \
1      ca1c90897ce1f12c092fe96acfc8832c  7d3278bd8d29df4dbb5dae11ed4e69bd   
3      981fb8da358abde5fc063be80e011720  7d3278bd8d29df4dbb5dae11ed4e69bd   
4      981fb8da358abde5fc063be80e011720  7d3278bd8d29df4dbb5dae11ed4e69bd   
12     3dcaa2ca32c9a3b8cdc566b259f754f0  7d3278bd8d29df4dbb5dae11ed4e69bd   
16     ad657d9e2b87df0576cb0f73e43a8049  7d3278bd8d29df4dbb5dae11ed4e69bd   
...                                 ...                               ...   
45901  9cf9888f491b2ee122be311428f3ae4a  7d3278bd8d29df4dbb5dae11ed4e69bd   
45909  42ac25c2d95a476f0186fa454c3def84  7d3278bd8d29df4dbb5dae11ed4e69bd   
45910  1c347a2ffbbdfb70197160e8e77c9a49  7d3278bd8d29df4dbb5dae11ed4e69bd   
45912  42ac25c2d95a476f0186fa454c3def84  7d3278bd8d29df4dbb5dae11ed4e69bd   
45914  9cf9888f491b2ee122be311428f3ae4a  7d3278bd8d29df4dbb5dae11ed4e69bd   

                 STARTED_TIME           FINISHED_TIME  CHARGELEVELSTART  \
1     2023-01-15 08:12:02.460 2023-01-15 08:17:27.798               NaN   
3     2023-01-15 12:57:37.334 2023-01-15 12:59:16.264               NaN   
4     2023-01-15 13:00:56.392 2023-01-15 13:04:52.894               NaN   
12    2023-01-17 09:35:05.107 2023-01-17 09:36:39.852               NaN   
16    2023-01-17 09:24:42.752 2023-01-17 09:47:33.661               NaN   
...                       ...                     ...               ...   
45901 2023-09-07 09:06:36.165 2023-09-07 09:12:54.068               NaN   
45909 2023-09-08 11:29:06.536 2023-09-08 11:40:29.025               NaN   
45910 2023-09-08 11:41:04.092 2023-09-08 11:45:55.177               NaN   
45912 2023-09-08 13:18:46.620 2023-09-08 13:20:49.256               NaN   
45914 2023-09-10 01:11:54.272 2023-09-10 01:28:32.977               NaN   

       CHARGELEVELEND  CHARGED  SERVICERENTAL               DURATION  \
1                65.0    False           True 0 days 00:05:25.338000   
3               100.0    False           True 0 days 00:01:38.930000   
4                98.0    False           True 0 days 00:03:56.502000   
12                3.0     True           True 0 days 00:01:34.745000   
16                2.0    False           True 0 days 00:22:50.909000   
...               ...      ...            ...                    ...   
45901            16.0     True           True 0 days 00:06:17.903000   
45909            40.0     True           True 0 days 00:11:22.489000   
45910            59.0     True           True 0 days 00:04:51.085000   
45912           100.0     True           True 0 days 00:02:02.636000   
45914            98.0    False           True 0 days 00:16:38.705000   

       DURATION_MIN  
1          5.422300  
3          1.648833  
4          3.941700  
12         1.579083  
16        22.848483  
...             ...  
45901      6.298383  
45909     11.374817  
45910      4.851417  
45912      2.043933  
45914     16.645083  

[5737 rows x 10 columns]

**Null values in charge level start : **
Correlation with missing charge level start data and service rental operation.


In [36]:
df_cleaned[df_cleaned['CHARGED'].isna()]

VEHICLE_ID                          MODEL_ID  \
6550   3edacefb15e7dbd9c470616287fb15ed  7d3278bd8d29df4dbb5dae11ed4e69bd   
6551   8434d49aa0baa7dad7a1df72defa440a  7d3278bd8d29df4dbb5dae11ed4e69bd   
6552   0c5d1b8f3dbe2f559fd3b70a6a31a0ce  7d3278bd8d29df4dbb5dae11ed4e69bd   
6578   8bf2cf4a2196a436618aa24ac35af4b8  7d3278bd8d29df4dbb5dae11ed4e69bd   
15166  b0021abffae86782e17f773e74fbe4a3  7d3278bd8d29df4dbb5dae11ed4e69bd   
15167  ad657d9e2b87df0576cb0f73e43a8049  7d3278bd8d29df4dbb5dae11ed4e69bd   
15190  1e3398251d2ef4394fdfd399bff97c56  7d3278bd8d29df4dbb5dae11ed4e69bd   
26124  18a86bed65d052f506e5dedcda224325  7d3278bd8d29df4dbb5dae11ed4e69bd   
32654  73f583278b1fb9f19621c240df5b31c5  7d3278bd8d29df4dbb5dae11ed4e69bd   
35444  61bedc330d875425f1eb7937a694cc21  7d3278bd8d29df4dbb5dae11ed4e69bd   
41716  d85d782906d7062168d02c28cdaca753  7d3278bd8d29df4dbb5dae11ed4e69bd   
43975  ca1c90897ce1f12c092fe96acfc8832c  7d3278bd8d29df4dbb5dae11ed4e69bd   

                 STARTED_TIME           FINISHED_TIME  CHARGELEVELSTART  \
6550  2023-04-04 13:41:24.942 2023-04-04 14:31:39.051               NaN   
6551  2023-04-04 14:00:34.126 2023-04-04 14:33:03.185               NaN   
6552  2023-04-04 14:14:38.530 2023-04-04 14:33:20.018               NaN   
6578  2023-04-04 13:10:07.228 2023-04-04 15:14:02.246               NaN   
15166 2023-04-04 09:22:46.116 2023-04-04 14:49:52.190               NaN   
15167 2023-04-04 14:00:54.721 2023-04-04 14:50:15.700               NaN   
15190 2023-04-04 11:51:25.360 2023-04-04 16:14:47.612               NaN   
26124 2023-04-04 14:28:27.021 2023-04-04 15:01:25.668               NaN   
32654 2023-04-04 13:59:20.945 2023-04-04 15:53:33.670               NaN   
35444 2023-04-04 14:22:29.324 2023-04-04 15:38:52.623               NaN   
41716 2023-04-04 13:40:04.145 2023-04-04 15:35:02.377               NaN   
43975 2023-04-04 11:14:44.686 2023-04-05 10:39:24.852               NaN   

       CHARGELEVELEND  CHARGED  SERVICERENTAL               DURATION  \
6550              NaN     <NA>           True 0 days 00:50:14.109000   
6551              NaN     <NA>          False 0 days 00:32:29.059000   
6552              NaN     <NA>          False 0 days 00:18:41.488000   
6578              NaN     <NA>          False 0 days 02:03:55.018000   
15166             NaN     <NA>          False 0 days 05:27:06.074000   
15167             NaN     <NA>          False 0 days 00:49:20.979000   
15190             NaN     <NA>          False 0 days 04:23:22.252000   
26124             NaN     <NA>           True 0 days 00:32:58.647000   
32654             NaN     <NA>          False 0 days 01:54:12.725000   
35444             NaN     <NA>          False 0 days 01:16:23.299000   
41716             NaN     <NA>          False 0 days 01:54:58.232000   
43975             NaN     <NA>          False 0 days 23:24:40.166000   

       DURATION_MIN  
6550      50.235150  
6551      32.484317  
6552      18.691467  
6578     123.916967  
15166    327.101233  
15167     49.349650  
15190    263.370867  
26124     32.977450  
32654    114.212083  
35444     76.388317  
41716    114.970533  
43975   1404.669433

Null values in 'Charged' : 12 lines with incomplete informations for the analysis. >> Excluding them from the df.

In [37]:
df_cleaned = df_cleaned[df_cleaned['CHARGED'].notna()]

In [39]:
df_cleaned[df_cleaned['CHARGELEVELEND'].isna()]

VEHICLE_ID                          MODEL_ID  \
186    ad5a12e5c0e12dea3c435931a78fa540  7d3278bd8d29df4dbb5dae11ed4e69bd   
237    ad5a12e5c0e12dea3c435931a78fa540  7d3278bd8d29df4dbb5dae11ed4e69bd   
341    ad5a12e5c0e12dea3c435931a78fa540  7d3278bd8d29df4dbb5dae11ed4e69bd   
621    fe24aa75d8c4cfa99df9e08546308ed6  7d3278bd8d29df4dbb5dae11ed4e69bd   
731    1ad57f261423114ec397fc6bcf156335  7d3278bd8d29df4dbb5dae11ed4e69bd   
...                                 ...                               ...   
44895  af77c8463ac27b3776ae4528d58704dc  7d3278bd8d29df4dbb5dae11ed4e69bd   
44966  0819698514bc1e0ddaaad0522259d4dd  7d3278bd8d29df4dbb5dae11ed4e69bd   
45524  308330becfa60294f45d93d1d359cc15  7d3278bd8d29df4dbb5dae11ed4e69bd   
45587  73f583278b1fb9f19621c240df5b31c5  7d3278bd8d29df4dbb5dae11ed4e69bd   
45727  99a39a0fcec88de0c6c067adb64a3bb6  7d3278bd8d29df4dbb5dae11ed4e69bd   

                 STARTED_TIME           FINISHED_TIME  CHARGELEVELSTART  \
186   2023-01-31 13:56:52.333 2023-01-31 14:09:25.769              85.0   
237   2023-01-28 09:38:55.152 2023-01-28 09:50:38.769              85.0   
341   2023-01-28 10:49:29.276 2023-01-28 10:51:20.769              74.0   
621   2023-02-24 08:42:39.776 2023-02-24 08:50:11.074              52.0   
731   2023-03-07 14:06:39.069 2023-03-07 14:27:16.145              90.0   
...                       ...                     ...               ...   
44895 2023-05-20 20:31:04.987 2023-05-20 20:42:29.640              87.0   
44966 2023-05-20 10:57:23.994 2023-05-20 11:17:12.769             100.0   
45524 2023-08-02 07:59:09.920 2023-08-02 08:07:58.142              99.0   
45587 2023-07-14 16:23:28.696 2023-07-14 17:46:02.833              97.0   
45727 2023-08-06 10:44:07.763 2023-08-06 11:37:42.769              54.0   

       CHARGELEVELEND  CHARGED  SERVICERENTAL               DURATION  \
186               NaN    False          False 0 days 00:12:33.436000   
237               NaN    False          False 0 days 00:11:43.617000   
341               NaN    False          False 0 days 00:01:51.493000   
621               NaN    False          False 0 days 00:07:31.298000   
731               NaN    False          False 0 days 00:20:37.076000   
...               ...      ...            ...                    ...   
44895             NaN    False          False 0 days 00:11:24.653000   
44966             NaN    False          False 0 days 00:19:48.775000   
45524             NaN     True          False 0 days 00:08:48.222000   
45587             NaN     True          False 0 days 01:22:34.137000   
45727             NaN     True          False 0 days 00:53:35.006000   

       DURATION_MIN  
186       12.557267  
237       11.726950  
341        1.858217  
621        7.521633  
731       20.617933  
...             ...  
44895     11.410883  
44966     19.812917  
45524      8.803700  
45587     82.568950  
45727     53.583433  

[246 rows x 10 columns]

**Null values in CHARGELEVELEND :** This 258 lines contain other informations needed for the analysis. No need to delete them from the global df.

In [52]:
df_cleaned['CHARGELEVELSTART'].value_counts()

CHARGELEVELSTART
100.0    1547
96.0      681
99.0      666
95.0      657
97.0      624
         ... 
14.0        8
11.0        5
10.0        5
13.0        4
0.0         3
Name: count, Length: 92, dtype: int64

**Exporting pre cleaned data for further analysis and aggregation with SQL**


In [48]:
df_cleaned.to_csv("free2move_cleaned.csv",index=False)